docs/osf/addons.rst

```diff
@@ -31,31 +32,228 @@ During your installation you created a virtual environment for OSF. Switch to th
+
+
+Addon Structure
+------------------
+
+An add-on SHOULD have the following folder structure
 
 ::
 
-    $ cookiecutter https://github.com/chrisseto/osf-Addon-cookiecutter.git
+    website/addons/addonshortname/
+    ├── __init__.py
+    ├── model.py
+    ├── requirements.txt
+    ├── routes.py
+    ├── settings
+    │   ├── __init__.py
+    │   └── defaults.py
+    ├── static
+    │   ├── comicon.png
+    │   ├── node-cfg.js*
+    │   ├── tests
+    │   │   └── ...
+    │   └── user-cfg.js*
+    ├── templates
+    │   ├── log_templates.mako
+    │   ├── addonshortname_node_settings.mako*
+    │   └── addonshortname_user_settings.mako*
+    ├── tests
+    │   ├── __init__.py
+    │   ├── test_model.py
+    │   └── test_views.py
+    └── views
+        └── ...
+
+\* optional
+
+StoredObject
+============
+
+A ``StoredObject`` is a class from our package ``modularodm`` that
+represents an object stored in a database.
+
+For our use case this database will always be TokuMX, a fork of mongodb
+that provides transactions and performance increases.
+
+Defining a ``StoredObject`` is very similar to the syntax used in
+`peewee`_ and `sqlalchemy`_
+
+The one caveat is that rather than using
+``modularodm.storedobject.StoredObject`` the base class must be
+``framework.mongo.StoredObject``
+
+Routes
+======
+
+A route is dictionary that containing one or more ``Rule`` objects in
+the ``'rules'`` key.
+
+Our url templating works the same way that `flask’s`_ does.
+
+.. code:: python
+
+    my_route = {
+      'rules': [
+        Rule(
+          [
+            '/my/<templated>/path/',  # Note all routes SHOULD end with a forward slash (/)
+            '/also/my/<templated>/path/'
+          ],
+          ('get', 'post'),  # Valid HTTP methods
+          view.my_view_function,  # The view method this route maps to
+          json_renderer  # The renderer used for this view function, either OsfWebRenderer or json_renderer
+        )
+      ]
+    }
+
+Routes SHOULD be defined in ``website.addons.youraddon.routes`` but
+could be defined anywhere
+
+Views
+=====
+
+Our views are implemented the same way that `flask’s`_ are.
+
+Any value matched by url templating (``<value_name>``) will be passed to
+your view function as a keyword argument
+
+Our framework supplies many python decorators to make writing view
+functions more pleasant.
+
+Below are a few examples that are commonly used in our code base.
+
+More can be found in ``website.project.decorators``.
+
+``framework.auth.decorators.must_be_logged_in``
+-----------------------------------------------
+
+Ensures that a user is logged in and imputes ``auth`` into keyword
+arguments
+
+``from website.project.decorators.must_have_addon``
+---------------------------------------------------
+
+``must_have_addon`` is a decorator factory meaning you must supply
+arguments to it to get a decorator.
+
+.. code:: python
+
+    def must_have_addon(addon_short_name, model_name):
```

Comment:
It makes it look like devs are supposed to implement this function themselves. Better to show the import and 1 or 2 usage examples, as you have below.